### DISCUSSION TITLE
Semantic versioning of MOM6 and MOM_interface

### DISCUSSION THREAD

**Discussion by alperaltuntas** [NCAR/MOM6#210] [Ideas]
Currently, we don't have a convenient way of determining whether specific tags of MOM6 and MOM_interface are compatible. The most frequent cause of incompatibility between these two is changing file trees when we add new MOM6 modules. So I propose we adopt a synchronous tagging convention for both MOM6 and MOM_interface that is based on semantic versioning such that having the same major version number for both MOM6 and MOM_interface guarantees compatible APIs. 

Semantic versioning is described here: https://semver.org/

Example tags of our current date-based convention:
- dev/ncar_220106
- mi_220107

Possible examples of proposed semantic versioning:
- dev/ncar_2.3.45
- main_2.4.56

I am also debating abandoning the ChangeLog that we keep in MOM_interface. 

Thoughts? @gustavo-marques @mnlevy1981 @klindsay28

**Comment by mnlevy1981** [comment]
I like the idea of using semantic  versioning, but am unfamiliar with strategies for versioning two packages that interact with each other. Would we just guarantee that as long as the major version number of the tag for both packages match, you could use those two tags together? E.g. dev/ncar_2.3.45 and main_2.4.56 could be used together but dev/ncar_2.3.45 and main_3.0.0 might not be compatible?

**Comment by alperaltuntas** [comment]
> Would we just guarantee that as long as the major version number of the tag for both packages match, you could use those two tags together

Maybe not a broad guarantee of "you can use those two tags together", but a guarantee of MOM6 and MOM_interface APIs being compatible. In other words, one should be able to compile and run `dev/ncar_2.3.45` and `main_2.4.56` just fine from a software API standpoint. Whether they *can* be run together has both software and scientific aspects though. Having the same major versions might not necessarily guarantee scientific correctness or a scientifically approved configuration.

> dev/ncar_2.3.45 and main_3.0.0 might not be compatible?

Yes, indeed `ncar_2.3.45` and `main_3.0.0` would be *guaranteed* to be incompatible. The users wouldn't be able to compile or run the model, as those tags would certainly have incompatible APIs.

**Comment by marshallward** [comment]
Would versioning of mom-ocean/MOM6 be useful here?  Or is that separate from the tagging of NCAR/MOM6?  Should we raise this at a future MOM meeting?

**Comment by alperaltuntas** [comment]
@marshallward, I'd be happy to participate in a future discussion of versioning mom-ocean/MOM6, but I guess this is separate from that. I was thinking more along the lines of establishing a synchronized versioning of our MOM6 dev/ncar branch and NCAR's superstructure repository called MOM_interface. We have a hierarchy of CESM tag specification: `CESM -> MOM_interface -> MOM6`. So this is not an issue for those that work with standard CESM development and release tags, but for those (like me) who may carry out separate developments on multiple branches of MOM6 and/or MOM_interface, things tend to diverge.

**Reply by marshallward** [reply]
Ok, no worries then, let's not worry about it until necessary.
